apps/emqx/src/emqx_router.erl

```diff
@@ -679,31 +681,89 @@ deinit_schema() ->
+                {Node, v1, legacy};
+            (Node, {error, {exception, badarg, _Stacktrace}}) ->
+                %% Likely, persistent term is not defined yet.
+                {Node, unknown, starting};
+            (Node, Error) ->
+                {Node, unknown, Error}
+        end,
+        Nodes,
+        emqx_router_proto_v1:get_routing_schema_vsn(Nodes)
+    ),
+    case lists:usort([Vsn || {_Node, Vsn, _} <- Responses, Vsn /= unknown]) of
+        [Vsn] when Vsn =:= v1; Vsn =:= v2 ->
+            Vsn;
+        [] ->
+            ?SLOG(notice, #{
+                msg => "cluster_routing_schema_discovery_failed",
+                responses => Responses,
+                reason =>
+                    "Could not determine configured routing storage schema in the cluster."
+            }),
+            undefined;
+        [_ | _] ->
+            ?SLOG(critical, #{
+                msg => "conflicting_routing_schemas_configured_in_cluster",
+                responses => Responses,
+                reason =>
+                    "There are nodes in the cluster with different configured routing "
+                    "storage schemas. This probably means that some nodes use v1 schema "
+                    "and some use v2, independently of each other. The routing is likely "
+                    "broken. Manual intervention required."
+            }),
+            error(conflicting_routing_schemas_configured_in_cluster)
+    end.
+
+-spec choose_schema_vsn(schemavsn(), schemavsn() | undefined) -> schemavsn().
+choose_schema_vsn(ConfSchema, ClusterSchema) ->
+    case detect_table_schema_vsn() of
+        [ClusterSchema] ->
+            %% Table contents match configured schema in the cluster.
+            ClusterSchema;
+        [Schema] when ClusterSchema =:= undefined ->
+            %% There are existing records following some schema, we have to use it.
+            Schema;
+        [] ->
+            %% No records in the tables, use schema configured in the cluster if any,
+            %% otherwise use configured.
+            emqx_maybe:define(ClusterSchema, ConfSchema);
+        ConlictingSchemas ->
             ?SLOG(critical, #{
                 msg => "conflicting_routing_schemas_detected_in_cluster",
-                configured => ConfType,
+                detected => ConlictingSchemas,
+                configured => ConfSchema,
+                configured_in_cluster => ClusterSchema,
                 reason =>
-                    "There are records in the routing tables related to both v1 "
-                    "and v2 storage schemas. This probably means that some nodes "
-                    "in the cluster use v1 schema and some use v2, independently "
-                    "of each other. The routing is likely broken. Manual intervention "
-                    "and full cluster restart is required. This node will shut down."
+                    "There are records in the routing tables either related to both v1 "
```

Comment:
need to log which nodes have v1 routes inserted.
then instruct to stop listeners on these nodes and restart.